server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java

```diff
@@ -354,156 +430,170 @@ public ClusterMetadataManifest writeIncrementalMetadata(
+                results.put(uploadedMetadata.getComponent(), uploadedMetadata);
             }, ex -> {
-                assert ex instanceof RemoteStateTransferException;
                 logger.error(
-                    () -> new ParameterizedMessage("Exception during transfer of IndexMetadata to Remote {}", ex.getMessage()),
+                    () -> new ParameterizedMessage("Exception during transfer of Metadata Fragment to Remote {}", ex.getMessage()),
                     ex
                 );
                 exceptionList.add(ex);
             }),
             latch
         );
 
-        for (IndexMetadata indexMetadata : toUpload) {
-            // 123456789012_test-cluster/cluster-state/dsgYj10Nkso7/index/ftqsCnn9TgOX/metadata_4_1690947200
-            writeIndexMetadataAsync(clusterState, indexMetadata, latchedActionListener);
+        if (uploadSettingsMetadata) {
+            uploadTasks.add(
+                getAsyncMetadataWriteAction(
+                    clusterState,
+                    SETTING_METADATA,
+                    SETTINGS_METADATA_FORMAT,
+                    clusterState.metadata().persistentSettings(),
+                    listener
+                )
+            );
+        }
+        if (uploadCoordinationMetadata) {
+            uploadTasks.add(
+                getAsyncMetadataWriteAction(
+                    clusterState,
+                    COORDINATION_METADATA,
+                    COORDINATION_METADATA_FORMAT,
+                    clusterState.metadata().coordinationMetadata(),
+                    listener
+                )
+            );
+        }
+        if (uploadTemplateMetadata) {
+            uploadTasks.add(
+                getAsyncMetadataWriteAction(
+                    clusterState,
+                    TEMPLATES_METADATA,
+                    TEMPLATES_METADATA_FORMAT,
+                    clusterState.metadata().templatesMetadata(),
+                    listener
+                )
+            );
+        }
+        customToUpload.forEach(
+            (key, value) -> uploadTasks.add(
+                getAsyncMetadataWriteAction(
+                    clusterState,
+                    String.join(DELIMITER, CUSTOM_METADATA, key),
+                    CUSTOM_METADATA_FORMAT,
+                    value,
+                    listener
+                )
+            )
+        );
+        indexToUpload.forEach(indexMetadata -> { uploadTasks.add(getIndexMetadataAsyncAction(clusterState, indexMetadata, listener)); });
+
+        // start async upload of all required metadata files
+        for (CheckedRunnable<IOException> uploadTask : uploadTasks) {
+            uploadTask.run();
         }
 
         try {
-            if (latch.await(getIndexMetadataUploadTimeout().millis(), TimeUnit.MILLISECONDS) == false) {
+            if (latch.await(getGlobalMetadataUploadTimeout().millis(), TimeUnit.MILLISECONDS) == false) {
```

Comment:
Yeah, we can discuss this further because as now we only need one setting for upload timeout as we are uploading all the files parallely now. Probably add a new setting with a different name for this?

Reply:
These `latch.await` with timeout can be dangerous if not handled properly. For example, even if latch times out, the upload is still in progress, right? Can this result in same metadata getting uploaded multiple times?

We have seen this issue in translog upload.

Reply:
When the latch times out, it might be possible that metadata upload is still in progress. But, here the behaviour will be that master will step down and a new master will be elected which will not upload the same metadata as the term would be updated in that case.

Reply:
Also, these timeouts are dynamically configurable